FirebaseAppCheck/Sources/AppAttestProvider/API/FIRAppAttestAPIService.m

```diff
@@ -56,8 +57,30 @@ - (instancetype)initWithAPIService:(id<FIRAppCheckAPIServiceProtocol>)APIService
 }
 
 - (nonnull FBLPromise<NSData *> *)getRandomChallenge {
-  // TODO: Implement.
-  return [FBLPromise resolvedWith:nil];
+  NSString *URLString =
+      [NSString stringWithFormat:@"%@/projects/%@/apps/%@:generateAppAttestChallenge",
+                                 self.APIService.baseURL, self.projectID, self.appID];
+  NSURL *URL = [NSURL URLWithString:URLString];
+
+  return [FBLPromise onQueue:[self defaultQueue]
+                          do:^id _Nullable {
+                            return [self.APIService sendRequestWithURL:URL
+                                                            HTTPMethod:@"POST"
+                                                                  body:nil
+                                                     additionalHeaders:nil];
+                          }]
+      .then(^id _Nullable(GULURLSessionDataResponse *_Nullable response) {
+        return [self randomChallengeWithAPIResponse:response];
+      });
+}
+
+- (FBLPromise<NSData *> *)randomChallengeWithAPIResponse:(GULURLSessionDataResponse *)response {
+  NSData *randomChallenge = response.HTTPBody;
+  return [FBLPromise resolvedWith:randomChallenge];
+}
```

Comment:
Also: I wasn't sure if this helper method should belong in this class or `FIRAppCheckAPIService` since that class does have a similar `tokenWithAPIResponse` helper method. On one hand, it involves manipulating a response from `self.APIService` (a `FIRAppCheckAPIService` instance) but on the other it seems more specific to App Attest than App Check.

Thoughts?

Reply:
`tokenWithAPIResponse` was moved up from specific API services to the `FIRAppCheckAPIService` because all FAC token exchange endpoints share the same response format, so it's better to share the implementation and having it in `FIRAppCheckAPIService` was the simplest approach.

As for parsing get challenge response - it should be in `FIRAppAttestAPIService.m` as it is unique for App Attest provider.

Reply:
Noted, thanks!